subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/VariantAttributeMatchingCache.java

```diff
@@ -117,4 +107,41 @@ private boolean matchAttributes(AttributeContainerInternal actual, AttributeCont
         private final Map<AttributeContainer, Boolean> ignoreExtraActual = Maps.newConcurrentMap();
         private final Map<AttributeContainer, ConsumerVariantMatchResult> transforms = Maps.newConcurrentMap();
     }
+
+    private static class ChainedTransformer implements ArtifactTransformer {
+        private final ConsumerVariantMatchResult.ConsumerVariant inputVariant;
+        private final VariantTransformRegistry.Registration candidate;
+
+        public ChainedTransformer(ConsumerVariantMatchResult.ConsumerVariant inputVariant, VariantTransformRegistry.Registration candidate) {
+            this.inputVariant = inputVariant;
+            this.candidate = candidate;
+        }
+
+        @Override
+        public List<File> transform(File file) {
+            List<File> result = new ArrayList<File>();
+            for (File intermediate : inputVariant.transformer.transform(file)) {
+                result.addAll(candidate.getArtifactTransform().transform(intermediate));
+            }
+            return result;
+        }
+
+        @Override
+        public boolean hasCachedResult(File input) {
+            if (inputVariant.transformer.hasCachedResult(input)) {
+                for (File intermediate : inputVariant.transformer.transform(input)) {
+                    if (!candidate.getArtifactTransform().hasCachedResult(intermediate)) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            return false;
+        }
```

Comment:
:o: I think we do not have any test coverage for `hasCachedResult()`. Since it needs to be in sync with what `transform()` does we should add some. Otherwise it would be easy to break. Since `transform()` is always called anyway; independent of the `hasCachedResult ()` result.
It could be unit test coverage for the two implementations of  `transform()` / `hasCachedResult()` and/or some coverage for the build operation handling in `ConsumerProvidedResolvedVariant` (see other comment).